ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,312 @@
+                        break;
+                    }
+                }
+            }
+
+            ++process_id;
+        }
+    }
+}
+
+void phreeqcIO::writeInputsToFile()
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _input_file);
+    std::ofstream outFile(_input_file, std::ofstream::out);
+
+    if (!outFile.good())
+        ERR("Could not open file %s.", _input_file);
+
+    outFile << *this;
+    outFile.close();
+}
+
+std::ofstream& operator<<(std::ofstream& outFile, phreeqcIO const& phreeqcIO)
+{
+    int const nNodes = phreeqcIO._solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        auto const& solution = phreeqcIO._solutions[node_id];
+        outFile << "SOLUTION " << node_id + 1 << "\n";
+        outFile << solution << "\n";
+
+        auto const& nodal_equilibrium_phases =
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            outFile << "EQUILIBRIUM_PHASES " << node_id + 1 << "\n";
+            outFile << nodal_equilibrium_phases << "\n";
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            outFile << "KINETICS " << node_id + 1 << "\n";
+            outFile << nodal_kinetic_reactant;
+            outFile << "-steps " << phreeqcIO._dt << "\n" << "\n";
+        }
+
+        if (node_id == 0)
+        {
+            auto const& reaction_rates = phreeqcIO._reaction_rates;
+            if (!reaction_rates.empty())
+            {
+                outFile << "RATES" << "\n";
+                outFile << reaction_rates << "\n";
+            }
+
+            outFile << "SELECTED_OUTPUT" << "\n";
+            outFile << *phreeqcIO._output;
+        }
+
+        outFile << "END" << "\n" << "\n";
+    }
+
+    return outFile;
+}
+
+void phreeqcIO::execute()
+{
+    auto database_loaded = [&](int instance_id) {
+        if (LoadDatabase(instance_id, _database.c_str()) == 0)
+            return true;
+        else
+            return false;
+    };
+
+    DBUG("Phreeqc: Executing chemical calculation.");
+    // initialize phreeqc configurations
+    auto const instance_id = CreateIPhreeqc();
+
+    // load a specific database in the working directory
+    if (!database_loaded(instance_id))
+        ERR("Failed to load the specified thermodynamic database file %s.",
+            _database.c_str());
+
+    SetSelectedOutputFileOn(instance_id, 1);
+
+    if (RunFile(instance_id, _input_file) > 0)
+        OutputErrorString(instance_id);
+}
+
+void phreeqcIO::readOutputsFromFile()
+{
+    auto const& basic_output_setups = _output->basic_output_setups;
+    auto const& output_file = basic_output_setups.getOutputFile();
+    DBUG("Reading phreeqc outputs from file '%s'.", output_file);
+    std::ifstream inFile(output_file, std::ifstream::in);
```

Comment:
This is highly confusing: an input stream from output file?

Reply:
the `ifstream` already has the `in` mode set: https://en.cppreference.com/w/cpp/io/basic_ifstream/basic_ifstream

Reply:
Fixed.

Reply:
> This is highly confusing: an input stream from output file?

Somewhat confusing. The object inFile is to the entities in ogs, while output_file is to phreeqc. Do you have good idea for improvement?

Reply:
Maybe calling them `iphreeqc_input_file` and `iphreeqc_result_file`?

Reply:
Renamed as recommended.